login-module/src/core.js

```diff
@@ -0,0 +1,77 @@
+import constants from './constants'
+import { WebAuth } from 'auth0-js'
+import Quiz from './quiz'
+import Enrollment from './enrollment'
+import Certificate from './certificate'
+
+export default class GraphAcademyCore {
+	constructor(options = {}) {
+		this.options = { ...constants.DEFAULT_OPTIONS, ...options }
+		const hasRequiredOptions = this.hasRequiredOptions(options)
+		if (!hasRequiredOptions) {
+			console.log(`required params missing - one of ${constants.REQUIRED_OPTIONS.join(', ')}`)
+			return
+		}
+		this.webAuth = new WebAuth({
+			clientID: 'hoNo6B00ckfAoFVzPTqzgBIJHFHDnHYu',
+			domain: 'login.neo4j.com',
+			redirectUri: `${window.location.origin}/accounts/login`,
+			audience: 'neo4j://accountinfo/',
+			scope: 'read:account-info openid email profile user_metadata',
+			responseType: 'token id_token'
+		})
+		this.certificate = new Certificate(this.options.trainingClassName, this.options.stage)
+		this.enrollment = new Enrollment(this.options.trainingClassName, this.options.stage)
+		this.quiz = new Quiz(this.options.trainingClassName, this.options.stage)
+	}
+
+	hasRequiredOptions(options) {
+		return constants.REQUIRED_OPTIONS.every(item => options[item])
+	}
+
+	async checkSession() {
+		const { options, webAuth } = this
+		return new Promise((resolve, reject) => {
+			webAuth.checkSession({}, async (err, result) => {
+				if (err) {
+					delete this.authResult
+					if (options.loginRedirectUrl) {
+						this.redirectToLogin()
+					}
+					return reject(err)
+				}
+				this.authResult = result
+				resolve(result)
+			})
+		})
+	}
+
+	async enrollStudentInClass (firstName, lastName) {
+		return await this.enrollment.enrollStudentInClass(firstName, lastName, await this.getAccessToken())
+	}
+
+	logout() {
+		const { options } = this
+		const logoutOptions = {}
+		if (options.logoutOptions && options.logoutOptions.shouldRedirect) {
+			logoutOptions.redirectTo = options.redirectOnLogout
+		}
+		this.webAuth.logout(logoutOptions)
+	}
+
+	redirectToLogin() {
+		const { options } = this
+		return window.location.href = options.loginRedirectUrl
+	}
+
+	async getAccessToken () {
+		let accessToken
+		if (this.authResult && this.authResult.accessToken) {
+			accessToken = this.authResult.accessToken
+		} else {
+			const authResult = await this.checkSession()
+			accessToken = authResult.accessToken
```

Comment:
Let's not have `checkSession` here. `getAccessToken` should only be called after `checkSession` done.

Reply:
> getAccessToken should only be called after checkSession done.

What happens if you don't? Should we return null / undefined? Then what happens if you call `enrollStudentInClass`?

```js
const graphAcademy = new GraphAcademy({})
graphAcademy.enrollStudentInClass('Guillaume', 'Grossetie') 
```

If we want to do that, then `enrollStudentInClass` should not be available on `GraphAcademy` but on a "GraphAcademy authenticated object":

```js
const graphAcademy = new GraphAcademy({})
// graphAcademy.enrollStudentInClass() is not available
const graphAcademyAuthContext = graphAcademy.checkSession() // should be renamed to getAuthContext() or something
graphAcademyAuthContext.enrollStudentInClass() // ok!
```

My point is that we don't want to create invalid states or require a particular call order because that's just bad design.

Reply:
I like the idea of moving enrollment functions to the "GraphAcademy authenticated object" context. And it makes sense because it should only be available on the authenticated user. In fact even the quiz and certificate related functions should move to authenticated object context since all these are only available and called on the authenticated user.

Reply:
Yes, exactly!